elkscmd/sash/sash.1
.TH SASH 1
.SH NAME
sash \- stand-alone shell with built-in commands
.SH SYNOPSYS
.B sash
.SH DESCRIPTION
The
.B sash
program is a stand-alone shell which is useful for recovering from certain
types of system failures.  In particular, it was created in order to cope
with the problem of missing shared libraries.
You can also use
.B sash
to safely upgrade to new versions of the shared libraries.
.PP
.B Sash
can execute external programs, as in any shell.  There are no restrictions
on these commands, as the standard shell is used to execute them if there
are any meta-characters in the command.
.PP
More importantly, however, is that many of the standard system commands
are built-in to
.BR sash .
These built-in commands are:
.PP
.NF
     -chgrp, -chmod, -chown, -cmp, -cp, -dd, -echo,
     -ed, -grep, -kill, -ln, -ls, -mkdir, -mknod,
     -more, -mount, -mv, -printenv, -pwd, -rm, -rmdir,
     -sync, -tar, -touch, -umount
.FI
.PP
These commands are generally similar to the standard programs with similar
names.  However, they are simpler and cruder than the external programs,
and so many of the options are not implemented.  The restrictions for each
built-in command are described later.
.PP
The built-in commands which correspond to external programs begin with a
dash character in order to distinguish them from the external programs.
So typing "ls", for example, will attempt to run the real
.B ls
program.
If "-ls" is typed, then the built-in command which mimics
.B ls
is called.
.PP
For the built-in commands, filenames are expanded so that asterisks,
question marks, and characters inside of square brackets are recognised
and are expanded.  However, no other command line processing is performed.
This includes quoting of arguments, specifying of file redirection, and
the specifying of a pipeline.
.PP
If an external program is non-existant or fails to run correctly, then
the "alias" built-in command may be used to redefine the standard command
so that it automatically runs the built-in command instead.  For example,
the command "alias ls -ls" redefines "ls" to run the built-in command.
This saves you the pain of having to remember to type the leading dash
all of the time.
.PP
The "help" command will list all of the built-in commands in
.B sash .
Each built-in command is described below in more detail.
.PP
.TP
.B alias [name [command]]
If
.I name
and
.I command
are provided, this defines an alias for a command
with the specified name, which executes the specified command, with
possible arguments.  If just
.I name
is provided, then the definition
of the specified command alias is displayed.  If nothing is provided,
then the definitions of all aliases are displayed.  When defining an
alias, wildcards are not expanded.
.TP
.B cd [dirname]
If
.I dirname
is provided, then the current directory is changed to the
dirname.  If
.I dirname
is absent, then the current directory is changed
to the user's home directory (value of the $HOME environment variable).
.TP
.B -chgrp gid filename ...
Change the group id for the specified list of files.  The
.I gid
can
either be a group name, or a decimal value.
.TP
.B -chmod mode filename ...
Change the mode of the specified list of files.  The
.I mode
argument
can only be an octal value.
.TP
.B -chown uid filename ...
Change the owner id for the specified list of files.  The
.I uid
can
either be a user name, or a decimal value.
.TP
.B -cmp filename1 filename2
Determines whether or not the specified filenames have identical data.
This says that the files are links to each other, are different sizes,
differ at a particular byte number, or are identical.
.TP
.B -cp srcname ... destname
Copies one or more files from the
.I srcname
to the
.IR destname .
If more
than one srcname is given, or if destname is a directory, then all
the srcnames are copied into the destname directory with the same
names as the srcnames.
.TP
.B -dd if=name of=name [bs=n] [count=n] [skip=n] [seek=n]
Copy data from one file to another with the specified parameters.
The
.I if
and
.I of
arguments must be provided, so stdin and stdout cannot
be specified.  The
.I bs
argument is the block size, and is a numeric
value (which defaults to 512 bytes).
.I Count
is the number of blocks
to be copied (which defaults to end of file for the input file).
.I Skip
is the number of blocks to ignore before copying (seek is used
if possible, and the default is 0).
.I Seek
is the number of blocks to
seek in the output file before writing (and defaults to 0).  Any of
the numeric decimal values can have one or more trailing letters
from the set 'kbw', which multiplies the value by 1024, 512, and 2
respectively.  The command reports the number of full blocks read
and written, and whether or not any partial block was read or written.
.TP
.B -echo [args] ...
Echo the arguments to the -echo command.  Wildcards are expanded, so
this is convenient to get a quick list of filenames in a directory.
The output is always terminated with a newline.
.TP
.B -ed [filename]
Edit the specified file using line-mode commands.  The following
.B ed
commands are provided: = c r w i a d p l s f k z and q.
Line numbers can be constants, ".", "$", "'x",
.RI / string /
and simple
arithmetic combinations of these.  The substitute command and the
search expression can only use literal strings.  There are some
small differences in the way that some commands behave.
.TP
.B exec filename [args]
Execute the specified program with the specified arguments.
This replaces
.B sash
completely by the executed program.
.TP
.B exit
Quit from
.BR sash .
.TP
.B -grep [-in] word filename ...
Display lines of the specified files which contain the given word.
If only one filename is given, then only the matching lines are
printed.  If multiple filenames are given, then the filenames are
printed along with the matching lines.
.I Word
must be a single word,
(ie, not a regular expression).  If -i is given, then case is
ignored when doing the search.  If -n is given, then the line
numbers of the matching lines are also printed.
.TP
.B help
Displays a list of built-in commands.
.TP
.B -kill [-signal] pid ...
Sends the specified signal to the specified list of processes.
.I Signal
is a numberic value, or one of the special values HUP, INT,
QUIT, or KILL.
.TP
.B -ln [-s] srcname ... destname
Links one or more files from the
.I srcname
to the specified
.IR destname .
If there are
multiple srcnames, or destname is a directory, then the link is
put in the destname directory with the same name as the source name.
The default links are hard links.  Using -s makes symbolic links.
For symbolic links, only one srcname can be specified.
.TP
.B -ls [-lid] filename ...
Display information about the specified filesnames, which may be
directories.  The normal listing is simply a list of filenames,
one per line.  The options available are -l, -i, and -d.  The -l
option produces a long listing given the normal 'ls' information.
The -i option also displays the inode numbers of the files.  The
-d option displays information about a directory, instead of the
files within it.
.TP
.B -mkdir dirname ...
Creates the specified directories.  They are created with the
default permissions.
.TP
.B -mknod filename type major minor
Creates a special device node, either a character file or a block
file.
.I Filename
is the name of the node.
.I Type
is either 'c' or 'd'.
.I Major
is the major device number.
.I Minor
is the minor device number.
Both of these numbers are decimal.
.TP
.B -more filename ...
Type out the contents of the specified filenames, one page at a
time.  For each page displayed, you can type 'n' and a return to go
to the next file, 'q' and a return to quit the command completely,
or just a return to go to the next page.
.TP
.B -mount [-t type] devname dirname
Mount a filesystem on a directory name.  The -t option specifies the
type of filesystem being mounted, and defaults to "minix".
.TP
.B -mv srcname ... destname
Moves one or more files from the
.I srcname
to the
.IR destname .
If multiple srcnames are given, or if destname is a directory, then
the srcnames are copied into the destination directory with the
same names as the srcnames.  Renames are attempted first, but if
this fails because of the files being on different filesystems,
then a copies and deletes are done instead.
.TP
.B -printenv [name]
If
.I name
is not given, this prints out the values of all the current
environment variables.  If
.I name
is given, then only that environment variable value is printed.
.TP
.B prompt [word] ...
Sets the prompt string that is displayed before reading of a
command.  A space is always added to the specified prompt.
.TP
.B -pwd
Prints the current working directory.
.TP
.B quit
Exits from
.BR sash .
.TP
.B -rm filename ...
Removes one or more files.
.TP
.B -rmdir dirname ...
Removes one or more directories.  The directories must be empty
for this to be successful.
.TP
.B setenv name value
Set the value of an environment variable.
.TP
.B source filename
Execute commands which are contained in the specified filename.
.TP
.B -sync
Do a "sync" system call to force dirty blocks out to the disk.
.TP
.B -tar [xtv]f devname [filename] ...
List or restore files from a tar archive.  This command can only
read tar files, not create them.  The available options are xtvf.
The f option must be specified, and accepts a device or file
name argument which contains the tar archive.  If no filename is
given, all files in the archive are listed or extracted.  Otherwise,
only those files starting with the specified filenames are done.
Leading slashes in the tar archive filenames are removed.
.TP
.B -touch filename ...
Updates the modify times of the specifed files.  If a file does not
exist, then it will be created with the default protection.
.TP
.B umask [mask]
If
.I mask
is given, sets the "umask" value used for initializing the
permissions of newly created files.  If
.I mask
is not given, then the
current umask value is printed.  The mask is an octal value.
.TP
.B -umount filename
Unmounts a file system.  The filename can either be the device name
which is mounted, or else the directory name which the file system
is mounted onto.
.TP
.B unalias name
Remove the definition for the specified alias.
.SH WARNINGS
.B Sash
should obviously be linked statically, otherwise it's purpose is lost.
.PP
The system is still vulnerable to unrunnable shared versions of
.B init
and
.B sh.
A patch to the kernel to run
.B sash
directly instead of
.B init
if the proper keyword is given to
.B lilo
would be very useful.
.PP
Several other system commands might be necessary for system recovery,
but aren't built-in to
.BR sash .
Among these are
.B uncompress
and
.BR fsck .
.SH AUTHOR
David I. Bell


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/sash/sash.c
/*
 * Copyright (c) 1993 by David I. Bell
 * Permission is granted to use, distribute, or modify this source,
 * provided that this copyright notice remains intact.
 *
 * Stand-alone shell for system maintainance for Linux.
 * This program should NOT be built using shared libraries.
 */

#include "sash.h"

#include <signal.h>
#include <errno.h>

typedef struct {
	char	*name;
	char	*usage;
	void	(*func)();
	int	minargs;
	int	maxargs;
} CMDTAB;


static	CMDTAB	cmdtab[] = {
#ifdef CMD_ALIAS
	"alias",	"[name [command]]", 	do_alias,
	1,		MAXARGS,
#endif

	"cd",		"[dirname]",		do_cd,
	1,		2,

#ifdef CMD_CHGRP
	"chgrp",	"gid filename ...",	do_chgrp,
	3,		MAXARGS,
#endif

#ifdef CMD_CHMOD
	"chmod",	"mode filename ...",	do_chmod,
	3,		MAXARGS,
#endif

#ifdef CMD_CHOWN
	"chown",	"uid filename ...",	do_chown,
	3,		MAXARGS,
#endif

#ifdef CMD_CMP
	"cmp",		"filename1 filename2",	do_cmp,
	3,		3,
#endif

#ifdef CMD_CP
	"cp",		"srcname ... destname",	do_cp,
	3,		MAXARGS,
#endif

#ifdef CMD_DD
	"dd",		"if=name of=name [bs=n] [count=n] [skip=n] [seek=n]", do_dd,
	3,		MAXARGS,
#endif

#ifdef CMD_ECHO
	"echo",	"[args] ...",		do_echo,
	1,		MAXARGS,
#endif

#ifdef CMD_ED
	"ed",		"[filename]",		do_ed,
	1,		2,
#endif

	"exec",		"filename [args]",	do_exec,
	2,		MAXARGS,

	"exit",		"",			do_exit,
	1,		1,

#ifdef CMD_GREP
	"grep",	"[-in] word filename ...",	do_grep,
	3,		MAXARGS,
#endif

#ifdef CMD_HELP
	"help",		"",			do_help,
	1,		MAXARGS,
#endif

#ifdef CMD_KILL
	"kill",	"[-sig] pid ...",	do_kill,
	2,		MAXARGS,
#endif

#ifdef CMD_LN
	"ln",		"[-s] srcname ... destname",	do_ln,
	3,		MAXARGS,
#endif

#ifdef CMD_LS
	"ls",		"[-lid] filename ...",	do_ls,
	1,		MAXARGS,
#endif

#ifdef CMD_MKDIR
	"mkdir",	"dirname ...",		do_mkdir,
	2,		MAXARGS,
#endif

#ifdef CMD_MKNOD
	"mknod",	"filename type major minor",	do_mknod,
	5,		5,
#endif

#ifdef CMD_MORE
	"more",	"filename ...",		do_more,
	2,		MAXARGS,
#endif

#ifdef CMD_MOUNT
	"mount",	"[-t type] devname dirname",	do_mount,
	3,		MAXARGS,
#endif

#ifdef CMD_MV
	"mv",		"srcname ... destname",	do_mv,
	3,		MAXARGS,
#endif

#ifdef CMD_PRINTENV
	"printenv",	"[name]",		do_printenv,
	1,		2,
#endif

#ifdef CMD_PROMPT
	"prompt",	"string",		do_prompt,
	2,		MAXARGS,
#endif

#ifdef CMD_PWD
	"pwd",		"",			do_pwd,
	1,		1,
#endif

	"quit",		"",			do_exit,
	1,		1,

#ifdef CMD_RM
	"rm",		"filename ...",		do_rm,
	2,		MAXARGS,
#endif

#ifdef CMD_RMDIR
	"rmdir",	"dirname ...",		do_rmdir,
	2,		MAXARGS,
#endif

#ifdef CMD_SETENV
	"setenv",	"name value",		do_setenv,
	2,		3,
#endif

#ifdef CMD_SOURCE
	"source",	"filename",		do_source,
	2,		2,
#endif

#ifdef CMD_SYNC
	"sync",	"",			do_sync,
	1,		1,
#endif

#ifdef CMD_TAR
	"tar",		"[xtv]f devname filename ...",	do_tar,
	2,		MAXARGS,
#endif

#ifdef CMD_TOUCH
	"touch",	"filename ...",		do_touch,
	2,		MAXARGS,
#endif

#ifdef CMD_UMASK
	"umask",	"[mask]",		do_umask,
	1,		2,
#endif

#ifdef CMD_MOUNT
	"umount",	"filename",		do_umount,
	2,		2,
#endif

#ifdef CMD_ALIAS
	"unalias",	"name",			do_unalias,
	2,		2,
#endif
	NULL,		0,			0,
	0,		0
};


#ifdef CMD_ALIAS
typedef struct {
	char	*name;
	char	*value;
} ALIAS;

static	ALIAS	*aliastable;
static	int	aliascount;
static	ALIAS	*findalias();
#endif

static	BOOL	intcrlf = TRUE;
static	char	*prompt;
static BOOL		isbinshell;

#ifdef CMD_SOURCE
static	FILE	*sourcefiles[MAXSOURCE];
static	int	sourcecount;
#endif

static	void	catchint();
static	void	catchquit();
static	void	readfile();
static	void	command();
static	void	runcmd();
static	void	showprompt();
static	BOOL	trybuiltin();

BOOL	intflag;


int main(int argc, char **argv)
{
	char	*cp;
	char	buf[PATHLEN];

	signal(SIGINT, catchint);
	signal(SIGQUIT, catchquit);
	signal(SIGTSTP, SIG_IGN);

	/* check if we are /bin/sh*/
	if ((cp = strrchr(argv[0], '/')) != 0)
		cp++;
	else cp = argv[0];
	isbinshell = !strcmp(cp, "sh");

#ifdef CMD_PROMPT
	if ((prompt = malloc(3)) == 0)
#endif
		prompt = "  ";
	strcpy(prompt, getuid()? "$ ": "# ");

	if (getenv("PATH") == NULL)
		putenv("PATH=/bin:/usr/bin:/sbin");

#ifdef CMD_SOURCE
#ifdef LATER
	cp = getenv("HOME");
	if (cp) {
		strcpy(buf, cp);
		strcat(buf, "/");
		strcat(buf, ".sashrc");

		if ((access(buf, 0) == 0) || (errno != ENOENT))
			readfile(buf);
	}
#endif

	if (argc > 1) {
		readfile(argv[1]);
	} else {
		readfile(NULL);
	}
#else
	readfile(NULL);
#endif

	exit(0);
}


/*
 * Read commands from the specified file.
 * A null name pointer indicates to read from stdin.
 */
static void
readfile(name)
	char	*name;
{
	FILE	*fp;
	int	cc;
	char	buf[CMDLEN];
#ifdef CMD_SOURCE
	BOOL	ttyflag;

	if (sourcecount >= MAXSOURCE) {
		fprintf(stderr, "Too many source files\n");
		return;
	}

	fp = stdin;
	if (name) {
		fp = fopen(name, "r");
		if (fp == NULL) {
			perror(name);
			return;
		}
	}
	sourcefiles[sourcecount++] = fp;

	ttyflag = isatty(fileno(fp));
#else
	fp = stdin;
#endif

	while (TRUE) {
		fflush(stdout);
		showprompt();

#ifdef CMD_SOURCE
		if (intflag && !ttyflag && (fp != stdin)) {
			fclose(fp);
			sourcecount--;
			return;
		}
#endif

		if (fgets(buf, CMDLEN - 1, fp) == NULL) {
			if (ferror(fp) && (errno == EINTR)) {
				clearerr(fp);
				continue;
			}
			break;
		}

		cc = strlen(buf);
		if (buf[cc - 1] == '\n')
			cc--;

		while ((cc > 0) && isblank(buf[cc - 1]))
			cc--;
		buf[cc] = '\0';

		command(buf);
	}

	if (ferror(fp)) {
		perror("Reading command line");
#ifdef CMD_SOURCE
		if (fp == stdin)
			exit(1);
#else
		exit(1);
#endif
	}

	clearerr(fp);
#ifdef CMD_SOURCE
	if (fp != stdin)
		fclose(fp);

	sourcecount--;
#endif
}

/* check if command line requires a real shell to run it*/
static BOOL
needfullshell(char *cmd)
{
	char 	*cp;

	for (cp = cmd; *cp; cp++) {
		if ((*cp >= 'a') && (*cp <= 'z'))
			continue;
		if ((*cp >= 'A') && (*cp <= 'Z'))
			continue;
		if (isdecimal(*cp))
			continue;
		if (isblank(*cp))
			continue;

		if ((*cp == '.') || (*cp == '/') || (*cp == '-') ||
			(*cp == '+') || (*cp == '=') || (*cp == '_') ||
			(*cp == ':') || (*cp == ',') || (*cp == '#'))
				continue;
#ifdef WILDCARDS
		if ((*cp == '*') || (*cp == '?') || (*cp == '[') || (*cp == ']'))
			continue;
#endif

		return TRUE;
	}
	return FALSE;
}


/*
 * Parse and execute one null-terminated command line string.
 * This breaks the command line up into words, checks to see if the
 * command is an alias, and expands wildcards.
 */
static void
command(char *cmd)
{
	char	**argv;
	int	argc;
#ifdef CMD_ALIAS
	ALIAS	*alias;
	char	buf[CMDLEN];
#endif

	intflag = FALSE;
#ifdef WILDCARDS
	freechunks();
#endif

	while (isblank(*cmd))
		cmd++;

	if ((*cmd == '\0') || (*cmd == '#') || !makeargs(cmd, &argc, &argv))
		return;

#ifdef CMD_ALIAS
	/*
	 * Search for the command in the alias table.
	 * If it is found, then replace the command name with
	 * the alias, and append any other arguments to it.
	 */
	alias = findalias(argv[0]);
	if (alias) {
		cmd = buf;
		strcpy(cmd, alias->value);

		while (--argc > 0) {
			strcat(cmd, " ");
			strcat(cmd, *++argv);
		}

		if (!makeargs(cmd, &argc, &argv))
			return;
	}
#endif

	/*
	 * Check if the command line has '>', '<', '&' etc that require a full shell.
	 * If so, then the external program will be run rather than the builtin.
	 */
	if (!needfullshell(cmd)) {
		if (trybuiltin(argc, argv))
			return;
	}

	/*
	 * Not a builtin or needs full shell, run the program along the PATH list.
	 */
	 runcmd(cmd, argc, argv);

}

/*
 * Try to execute a built-in command.
 * Returns TRUE if the command is a built in, whether or not the
 * command succeeds.  Returns FALSE if this is not a built-in command.
 */
static BOOL
trybuiltin(int argc, char **argv)
{
	CMDTAB	*cmdptr;
#ifdef WILDCARDS
	int	oac;
	int	matches;
	int	newargc;
	int	i;
	char	*newargv[MAXARGS];
	char	*nametable[MAXARGS];
#endif

	cmdptr = cmdtab - 1;
	do {
		cmdptr++;
		if (cmdptr->name == NULL)
			return FALSE;

	} while (strcmp(argv[0], cmdptr->name));

	/*
	 * Give a usage string if the number of arguments is too large
	 * or too small.
	 */
	if ((argc < cmdptr->minargs) || (argc > cmdptr->maxargs)) {
		fprintf(stderr, "usage: %s %s\n", cmdptr->name, cmdptr->usage);
		return TRUE;
	}

	/*
	 * Check here for several special commands which do not
	 * have wildcarding done for them.
	 */
	if (
#ifdef CMD_ALIAS
	(cmdptr->func == do_alias) ||
#endif
#ifdef CMD_PROMPT
	(cmdptr->func == do_prompt) ||
#endif
	   0) {
		(*cmdptr->func)(argc, argv);
		return TRUE;
	}

#ifdef WILDCARDS
	/*
	 * Now for each command argument, see if it is a wildcard, and if
	 * so, replace the argument with the list of matching filenames.
	 */
	newargv[0] = argv[0];
	newargc = 1;
	oac = 0;

	while (++oac < argc) {
		matches = expandwildcards(argv[oac], MAXARGS, nametable);
		if (matches < 0)
			return TRUE;

		if ((newargc + matches) >= MAXARGS) {
			fprintf(stderr, "Too many arguments\n");
			return TRUE;
		}

		if (matches == 0)
			newargv[newargc++] = argv[oac];

		for (i = 0; i < matches; i++)
			newargv[newargc++] = nametable[i];
	}

	(*cmdptr->func)(newargc, newargv);
#else
	(*cmdptr->func)(argc, argv);

#endif
	return TRUE;
}


/*
 * Execute the specified command.
 */
static void
runcmd(char *cmd, int argc, char **argv)
{
	int		pid;
	int		status;

	/*
	 * If a full shell is required, run 'sh -c cmd' unless we are the only shell.
	 */
	if (needfullshell(cmd)) {
		if (isbinshell)
			fprintf(stderr, "%s: No such file or directory\n", argv[0]);
		else {
			system(cmd);
			wait(&status);	/* synchronize shell prompt*/
		}
		return;
	}

	/*
	 * No magic characters in the command, so do the fork and
	 * exec ourself.  If this fails with ENOEXEC, then run the
	 * shell anyway since it might be a shell script.
	 */
	pid = vfork();

	if (pid < 0) {
		perror("fork failed");
		return;
	}

	if (pid) {
		status = 0;
		intcrlf = FALSE;

		while (((pid = wait(&status)) < 0) && (errno == EINTR))
			;

		intcrlf = TRUE;
		if ((status & 0xff) == 0)
			return;

		fprintf(stderr, "pid %d: %s (signal %d)\n", pid,
			(status & 0x80) ? "core dumped" : "killed",
			status & 0x7f);

		return;
	}

	/*
	 * We are the child, so run the program.
	 * First close any extra file descriptors we have opened.
	 */
#ifdef CMD_SOURCE
	while (--sourcecount >= 0) {
		if (sourcefiles[sourcecount] != stdin)
			fclose(sourcefiles[sourcecount]);
	}
#endif

	execvp(argv[0], argv);

	/*
	 * If file has access but didn't execute, guess it's a shell script
	 * and run 'sh -c cmd'.
	 */
	if (errno == ENOEXEC)
		execl("/bin/sh", "sh", "-c", cmd, (char*)0);

	perror(argv[0]);	/* Usually 'No such file or directory'*/
	exit(1);
}

#ifdef CMD_HELP
void
do_help(argc, argv)
	char	**argv;
{
	CMDTAB	*cmdptr;

	for (cmdptr = cmdtab; cmdptr->name; cmdptr++)
		printf("%-10s %s\n", cmdptr->name, cmdptr->usage);
}
#endif /* CMD_HELP */

#ifdef CMD_ALIAS
void
do_alias(argc, argv)
	char	**argv;
{
	char	*name;
	char	*value;
	ALIAS	*alias;
	int	count;
	char	buf[CMDLEN];

	if (argc < 2) {
		count = aliascount;
		for (alias = aliastable; count-- > 0; alias++)
			printf("%s\t%s\n", alias->name, alias->value);
		return;
	}

	name = argv[1];
	if (argc == 2) {
		alias = findalias(name);
		if (alias)
			printf("%s\n", alias->value);
		else
			fprintf(stderr, "Alias \"%s\" is not defined\n", name);
		return;
	}

	if (strcmp(name, "alias") == 0) {
		fprintf(stderr, "Cannot alias \"alias\"\n");
		return;
	}

	if (!makestring(argc - 2, argv + 2, buf, CMDLEN))
		return;

	value = malloc(strlen(buf) + 1);

	if (value == NULL) {
		fprintf(stderr, "No memory for alias value\n");
		return;
	}

	strcpy(value, buf);

	alias = findalias(name);
	if (alias) {
		free(alias->value);
		alias->value = value;
		return;
	}

	if ((aliascount % ALIASALLOC) == 0) {
		count = aliascount + ALIASALLOC;

		if (aliastable)
			alias = (ALIAS *) realloc(aliastable,
				sizeof(ALIAS *) * count);
		else
			alias = (ALIAS *) malloc(sizeof(ALIAS *) * count);

		if (alias == NULL) {
			free(value);
			fprintf(stderr, "No memory for alias table\n");
			return;
		}

		aliastable = alias;
	}

	alias = &aliastable[aliascount];

	alias->name = malloc(strlen(name) + 1);

	if (alias->name == NULL) {
		free(value);
		fprintf(stderr, "No memory for alias name\n");
		return;
	}

	strcpy(alias->name, name);
	alias->value = value;
	aliascount++;
}

/*
 * Look up an alias name, and return a pointer to it.
 * Returns NULL if the name does not exist.
 */
static ALIAS *
findalias(name)
	char	*name;
{
	ALIAS	*alias;
	int	count;

	count = aliascount;
	for (alias = aliastable; count-- > 0; alias++) {
		if (strcmp(name, alias->name) == 0)
			return alias;
	}

	return NULL;
}
#endif /* CMD_ALIAS */


#ifdef CMD_SOURCE
void
do_source(argc, argv)
	char	**argv;
{
	readfile(argv[1]);
}
#endif


void
do_exec(argc, argv)
	char	**argv;
{
	char	*name;

	name = argv[1];

	if (access(name, 4)) {
		perror(name);
		return;
	}

#ifdef CMD_SOURCE
	while (--sourcecount >= 0) {
		if (sourcefiles[sourcecount] != stdin)
			fclose(sourcefiles[sourcecount]);
	}
#endif

	argv[0] = name;
	argv[1] = NULL;

	execv(name, argv);

	perror(name);
	exit(1);
}

#ifdef CMD_PROMPT
void
do_prompt(argc, argv)
	char	**argv;
{
	char	*cp;
	char	buf[CMDLEN];

	if (!makestring(argc - 1, argv + 1, buf, CMDLEN))
		return;

	cp = malloc(strlen(buf) + 2);
	if (cp == NULL) {
		fprintf(stderr, "No memory for prompt\n");
		return;
	}

	strcpy(cp, buf);
	strcat(cp, " ");

	if (prompt)
		free(prompt);
	prompt = cp;
}
#endif

#ifdef CMD_ALIAS
void
do_unalias(argc, argv)
	char	**argv;
{
	ALIAS	*alias;

	while (--argc > 0) {
		alias = findalias(*++argv);
		if (alias == NULL)
			continue;

		free(alias->name);
		free(alias->value);
		aliascount--;
		alias->name = aliastable[aliascount].name;
		alias->value = aliastable[aliascount].value;
	}
}
#endif /* CMD_ALIAS */

/*
 * Display the prompt string.
 */
static void
showprompt()
{
	write(STDOUT, prompt, strlen(prompt));
}


static void
catchint()
{
	signal(SIGINT, catchint);

	intflag = TRUE;

	if (intcrlf)
		write(STDOUT, "\n", 1);
}


static void
catchquit()
{
	signal(SIGQUIT, catchquit);

	intflag = TRUE;

	if (intcrlf)
		write(STDOUT, "\n", 1);
}

/* END CODE */
